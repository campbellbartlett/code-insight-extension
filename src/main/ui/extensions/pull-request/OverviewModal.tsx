import React, { useCallback, useEffect, useState } from 'react';
import { ModalAction } from '@atlassian/clientside-extensions-components/lib/handlers/ModalHandler';
import { Context } from '@atlassian/clientside-extensions-registry/lib/types';
import { PullRequestInsightsContext } from './PullRequestInsightsContext';
import getCodeInsightExtensionsContext from './insight-context-api/get-insight-context';
import toggleAdminOverride from './insight-context-api/toggle-admin-override';
import { AdminOverrideStatus } from './admin-override-status';
import { ReportStatus } from './report-status/report-status';
import { OverviewAdmonition } from './overview-admonition';

const createMainButton = (
    isOverridden: boolean,
    loaded: boolean,
    userAdmin: boolean,
    onOverride: () => void
) => ({
    text: isOverridden ? 'Disable merge override' : 'Enable merging',
    onClick: () => {
        onOverride();
    },
    isDisabled: loaded && !userAdmin,
});

type ModalComponentProps = {
    context: Context<{}>;
    setActions: (actions: ModalAction[]) => void;
    createCloseButton: (refreshOnClose: boolean) => ModalAction;
    setIsOverride: (isOverride: boolean) => void;
};

const OverviewModal: React.FC<ModalComponentProps> = (props: ModalComponentProps) => {
    const { context, setActions, createCloseButton, setIsOverride } = props;
    const [insightsContext, setInsightsContext] = useState<PullRequestInsightsContext | null>(null);

    useEffect(() => {
        getCodeInsightExtensionsContext(context).then(
            (codeInsightContext: PullRequestInsightsContext) => {
                setInsightsContext(codeInsightContext);
                setActionButtons(true, codeInsightContext, false);
            }
        );
    }, [context]);

    const setActionButtons = useCallback(
        (isLoaded: boolean, insightContext: PullRequestInsightsContext, isUpdated: boolean) => {
            const isOverridden = insightContext && insightContext.adminOverride;
            const mainButton = createMainButton(
                isOverridden,
                isLoaded,
                isLoaded && insightContext && insightContext.userAdmin,
                async () => {
                    await toggleAdminOverride(context, insightContext);
                    const codeInsightContext = await getCodeInsightExtensionsContext(context);
                    setInsightsContext(codeInsightContext);
                    setActionButtons(true, codeInsightContext, true);
                }
            );
            const closeButton = createCloseButton(isUpdated);
            setIsOverride(isLoaded && insightContext && insightContext.adminOverride);
            setActions([mainButton, closeButton]);
        },
        [setIsOverride, setActions]
    );

    return (
        <div id="modal-div">
            {insightsContext && <AdminOverrideStatus isOverride={insightsContext.adminOverride} />}
            {insightsContext && (
                <ReportStatus codeInsightReports={insightsContext.codeInsightReports} />
            )}
            {insightsContext && <OverviewAdmonition userAdmin={insightsContext.userAdmin} />}
        </div>
    );
};

export default OverviewModal;
