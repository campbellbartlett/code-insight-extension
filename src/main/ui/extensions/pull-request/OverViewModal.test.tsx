import React from 'react';
import fetchMock from 'jest-fetch-mock';
import { act } from '@testing-library/react';
import { mount, ReactWrapper } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';
import { ModalAction } from '@atlassian/clientside-extensions-components/dist/handlers/ModalHandler';
import OverviewModal from './OverviewModal';

const context = {
    project: {
        key: 'projectKey',
    },
    repository: {
        slug: 'repoSlug',
    },
    pullRequest: {
        fromRef: {
            latestCommit: 'commitHash',
        },
    },
};

const mockReports = [
    {
        name: 'integrationtests',
        status: 'FAIL',
    },
    {
        name: 'blackduck',
        status: 'FAIL',
    },
    {
        name: 'sonar',
        status: 'PASS',
    },
];

const getMockReportResponse = (isUserAdmin: boolean, hasBeenOveridden: boolean) => {
    return `{
              "codeInsightReports": ${JSON.stringify(mockReports)},
              "commitHash": "edc172925cf340b88c499fe401f9597289142b8e",
              "repositorySlug": "docker-compose-for-bitbucket",
              "projectKey": "TP",
              "userAdmin": ${isUserAdmin},
              "adminOverride": ${hasBeenOveridden}
            }`;
};

let actions: ModalAction[] = [];
const setActions = jest.fn((newActions: ModalAction[]) => {
    actions = [...newActions];
});

const createCloseButton = jest.fn((hasBeenUpdated: boolean) => hasBeenUpdated);
const setIsAdminOverride = jest.fn();

async function mountModal() {
    const overviewModal = (
        <OverviewModal
            context={context}
            setActions={setActions}
            createCloseButton={createCloseButton}
            setIsOverride={setIsAdminOverride}
        />
    );

    let container: ReactWrapper;
    await act(async () => {
        container = mount(overviewModal);
    });
    container.update();
    return container;
}

describe('OverView modal renders correctly', () => {
    it('should render correctly', async (done: any) => {
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: getMockReportResponse(false, false),
            });
        });

        const container = await mountModal();

        expect(container).not.toBeNull();
        expect(mountToJson(container)).toMatchSnapshot();

        done();
    });

    it('should fetch and pass admin prop onto children correctly', async (done: any) => {
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: getMockReportResponse(true, false),
            });
        });

        const container = await mountModal();

        expect(container).not.toBeNull();
        expect(container.find('OverviewAdmonition').props().userAdmin).toBeTruthy();

        done();
    });

    it('should fetch and pass override prop onto children correctly', async (done: any) => {
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: getMockReportResponse(false, true),
            });
        });

        const container = await mountModal();

        expect(container).not.toBeNull();
        expect(container.find('AdminOverrideStatus').props().isOverride).toBeTruthy();

        done();
    });

    it('should fetch and pass reports prop onto children correctly', async (done: any) => {
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: getMockReportResponse(false, true),
            });
        });

        const container = await mountModal();

        expect(container).not.toBeNull();
        expect(container.find('ReportStatus').props().codeInsightReports).toEqual(mockReports);

        done();
    });

    it('should show the correct text on the override button when merging is disabled', async (done: any) => {
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: getMockReportResponse(true, false),
            });
        });

        const container = await mountModal();

        expect(container).not.toBeNull();
        expect(actions.length).toBe(2);
        // Expect one of the actions to be the button with text 'enable merging'
        expect(actions.some(action => action.text === 'Enable merging')).toBeTruthy();
        // Expect the other action to be the result of passing false into the createCloseButton mocked function
        expect(actions.some(action => action === false)).toBeTruthy();

        done();
    });

    it('should show the correct text on the override button when merging is enabled', async (done: any) => {
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: getMockReportResponse(true, true),
            });
        });

        const container = await mountModal();

        expect(container).not.toBeNull();
        expect(actions.length).toBe(2);
        // Expect one of the actions to be the button with text 'enable merging'
        expect(actions.some(action => action.text === 'Disable merge override')).toBeTruthy();
        // Expect the other action to be the result of passing false into the createCloseButton mocked function
        expect(actions.some(action => action === false)).toBeTruthy();

        done();
    });

    it('should show the override button as disabled when the user is not admin', async (done: any) => {
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: getMockReportResponse(false, true),
            });
        });

        const container = await mountModal();

        expect(container).not.toBeNull();
        expect(actions.length).toBe(2);
        // Expect that one of the action buttons is disabled
        expect(actions.find(action => action.isDisabled)).not.toBeNull();
        // Expect the other action to be the result of passing false into the createCloseButton mocked function
        expect(actions.some(action => action === false)).toBeTruthy();

        done();
    });

    it('should switch override button text and style when toggling admin authority to merge', async (done: any) => {
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: getMockReportResponse(true, true),
            });
        });

        const container = await mountModal();

        expect(container).not.toBeNull();
        expect(actions.some(action => action.text === 'Disable merge override')).toBeTruthy();

        // Mock the post request to update the admin override
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: {},
            });
        });

        // Mock the next fetch which gets the latest insight reports from the backend
        fetchMock.mockOnce(async () => {
            return Promise.resolve({
                body: getMockReportResponse(true, false),
            });
        });

        await act(async () => {
            // When performing the onClick function, update the status of the admin override
            // in the backend, refetch the latest reports and update the UI
            const button = actions.find(action => action.text === 'Disable merge override');
            expect(button).not.toBeNull();
            button.onClick();
        });

        expect(actions.length).toBe(2);
        // Expect that one of the action buttons is 'Enable Merging'
        expect(actions.some(action => action.text === 'Enable merging')).toBeTruthy();
        // Expect the other action to be the result of passing true into the createCloseButton mocked function
        expect(actions.some(action => action)).toBeTruthy();

        done();
    });
});
