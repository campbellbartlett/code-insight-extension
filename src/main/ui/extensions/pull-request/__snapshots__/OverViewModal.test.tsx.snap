// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverView modal renders correctly should render correctly 1`] = `
<OverviewModal
  context={
    Object {
      "project": Object {
        "key": "projectKey",
      },
      "pullRequest": Object {
        "fromRef": Object {
          "latestCommit": "commitHash",
        },
      },
      "repository": Object {
        "slug": "repoSlug",
      },
    }
  }
  createCloseButton={
    [MockFunction] {
      "calls": Array [
        Array [
          false,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": false,
        },
      ],
    }
  }
  setActions={
    [MockFunction] {
      "calls": Array [
        Array [
          Array [
            Object {
              "isDisabled": true,
              "onClick": [Function],
              "text": "Enable merging",
            },
            false,
          ],
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setIsOverride={
    [MockFunction] {
      "calls": Array [
        Array [
          false,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    id="modal-div"
  >
    <AdminOverrideStatus
      isOverride={false}
    >
      <div
        id="admin-override-status"
      >
        <h3
          id="standard-message"
        >
          This Pull Request cannot be merged until all code insight quality reports have passed.
        </h3>
      </div>
    </AdminOverrideStatus>
    <ReportStatus
      codeInsightReports={
        Array [
          Object {
            "name": "integrationtests",
            "status": "FAIL",
          },
          Object {
            "name": "blackduck",
            "status": "FAIL",
          },
          Object {
            "name": "sonar",
            "status": "PASS",
          },
        ]
      }
    >
      <div
        id="report-status"
        style={
          Object {
            "marginBottom": "1em",
            "marginTop": "1em",
          }
        }
      >
        <ul
          style={
            Object {
              "listStyleType": "none",
            }
          }
        >
          <li
            id="integrationtests"
            key="integrationtests"
          >
            <SingleReportStatus
              report={
                Object {
                  "name": "integrationtests",
                  "status": "FAIL",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "flex",
                    "justifyContent": "space-between",
                    "width": "10rem",
                  }
                }
              >
                <span
                  id="report-name"
                >
                  integrationtests
                </span>
                 
                <FailedStatus>
                  <StatusInParagraph
                    color="red"
                    text="Failed"
                  >
                    <div
                      style={
                        Object {
                          "width": "4em",
                        }
                      }
                    >
                      <WithAnalyticsEvents(StatusInternal)
                        color="red"
                        text="Failed"
                      >
                        <AnalyticsContextConsumer
                          createEventMap={
                            Object {
                              "onClick": [Function],
                              "onHover": [Function],
                            }
                          }
                          wrappedComponentProps={
                            Object {
                              "color": "red",
                              "text": "Failed",
                            }
                          }
                        >
                          <StatusInternal
                            color="red"
                            createAnalyticsEvent={[Function]}
                            onClick={[Function]}
                            onHover={[Function]}
                            text="Failed"
                          >
                            <span
                              className="status-lozenge-span"
                              data-color="red"
                              data-node-type="status"
                              onClick={[Function]}
                              onMouseEnter={[Function]}
                              onMouseLeave={[Function]}
                              title="Failed"
                            >
                              <Lozenge
                                appearance="removed"
                                isBold={false}
                                maxWidth={200}
                              >
                                <Provider>
                                  <Consumer
                                    appearance="removed"
                                    isBold={false}
                                    maxWidth={200}
                                  >
                                    <Component
                                      backgroundColor="#FFEBE6"
                                      maxWidth={200}
                                      textColor="#BF2600"
                                    >
                                      <EmotionCssPropInternal
                                        __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Component"
                                        __EMOTION_TYPE_PLEASE_DO_NOT_USE__="span"
                                        css={
                                          Object {
                                            "backgroundColor": "#FFEBE6",
                                            "borderRadius": "3px",
                                            "boxSizing": "border-box",
                                            "color": "#BF2600",
                                            "display": "inline-block",
                                            "fontSize": "11px",
                                            "fontWeight": 700,
                                            "lineHeight": 1,
                                            "maxWidth": "100%",
                                            "padding": "2px 0 3px 0",
                                            "textTransform": "uppercase",
                                            "verticalAlign": "baseline",
                                          }
                                        }
                                      >
                                        <span
                                          className="css-abx768-Component"
                                        >
                                          <Component
                                            backgroundColor="#FFEBE6"
                                            maxWidth={200}
                                            textColor="#BF2600"
                                          >
                                            <EmotionCssPropInternal
                                              __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Component"
                                              __EMOTION_TYPE_PLEASE_DO_NOT_USE__="span"
                                              css={
                                                Object {
                                                  "boxSizing": "border-box",
                                                  "display": "inline-block",
                                                  "maxWidth": "200px",
                                                  "overflow": "hidden",
                                                  "padding": "0 4px",
                                                  "textOverflow": "ellipsis",
                                                  "verticalAlign": "top",
                                                  "whiteSpace": "nowrap",
                                                  "width": "100%",
                                                }
                                              }
                                            >
                                              <span
                                                className="css-19vb09t-Component"
                                              >
                                                Failed
                                              </span>
                                            </EmotionCssPropInternal>
                                          </Component>
                                        </span>
                                      </EmotionCssPropInternal>
                                    </Component>
                                  </Consumer>
                                </Provider>
                              </Lozenge>
                            </span>
                          </StatusInternal>
                        </AnalyticsContextConsumer>
                      </WithAnalyticsEvents(StatusInternal)>
                    </div>
                  </StatusInParagraph>
                </FailedStatus>
              </div>
            </SingleReportStatus>
          </li>
          <li
            id="blackduck"
            key="blackduck"
          >
            <SingleReportStatus
              report={
                Object {
                  "name": "blackduck",
                  "status": "FAIL",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "flex",
                    "justifyContent": "space-between",
                    "width": "10rem",
                  }
                }
              >
                <span
                  id="report-name"
                >
                  blackduck
                </span>
                 
                <FailedStatus>
                  <StatusInParagraph
                    color="red"
                    text="Failed"
                  >
                    <div
                      style={
                        Object {
                          "width": "4em",
                        }
                      }
                    >
                      <WithAnalyticsEvents(StatusInternal)
                        color="red"
                        text="Failed"
                      >
                        <AnalyticsContextConsumer
                          createEventMap={
                            Object {
                              "onClick": [Function],
                              "onHover": [Function],
                            }
                          }
                          wrappedComponentProps={
                            Object {
                              "color": "red",
                              "text": "Failed",
                            }
                          }
                        >
                          <StatusInternal
                            color="red"
                            createAnalyticsEvent={[Function]}
                            onClick={[Function]}
                            onHover={[Function]}
                            text="Failed"
                          >
                            <span
                              className="status-lozenge-span"
                              data-color="red"
                              data-node-type="status"
                              onClick={[Function]}
                              onMouseEnter={[Function]}
                              onMouseLeave={[Function]}
                              title="Failed"
                            >
                              <Lozenge
                                appearance="removed"
                                isBold={false}
                                maxWidth={200}
                              >
                                <Provider>
                                  <Consumer
                                    appearance="removed"
                                    isBold={false}
                                    maxWidth={200}
                                  >
                                    <Component
                                      backgroundColor="#FFEBE6"
                                      maxWidth={200}
                                      textColor="#BF2600"
                                    >
                                      <EmotionCssPropInternal
                                        __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Component"
                                        __EMOTION_TYPE_PLEASE_DO_NOT_USE__="span"
                                        css={
                                          Object {
                                            "backgroundColor": "#FFEBE6",
                                            "borderRadius": "3px",
                                            "boxSizing": "border-box",
                                            "color": "#BF2600",
                                            "display": "inline-block",
                                            "fontSize": "11px",
                                            "fontWeight": 700,
                                            "lineHeight": 1,
                                            "maxWidth": "100%",
                                            "padding": "2px 0 3px 0",
                                            "textTransform": "uppercase",
                                            "verticalAlign": "baseline",
                                          }
                                        }
                                      >
                                        <span
                                          className="css-abx768-Component"
                                        >
                                          <Component
                                            backgroundColor="#FFEBE6"
                                            maxWidth={200}
                                            textColor="#BF2600"
                                          >
                                            <EmotionCssPropInternal
                                              __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Component"
                                              __EMOTION_TYPE_PLEASE_DO_NOT_USE__="span"
                                              css={
                                                Object {
                                                  "boxSizing": "border-box",
                                                  "display": "inline-block",
                                                  "maxWidth": "200px",
                                                  "overflow": "hidden",
                                                  "padding": "0 4px",
                                                  "textOverflow": "ellipsis",
                                                  "verticalAlign": "top",
                                                  "whiteSpace": "nowrap",
                                                  "width": "100%",
                                                }
                                              }
                                            >
                                              <span
                                                className="css-19vb09t-Component"
                                              >
                                                Failed
                                              </span>
                                            </EmotionCssPropInternal>
                                          </Component>
                                        </span>
                                      </EmotionCssPropInternal>
                                    </Component>
                                  </Consumer>
                                </Provider>
                              </Lozenge>
                            </span>
                          </StatusInternal>
                        </AnalyticsContextConsumer>
                      </WithAnalyticsEvents(StatusInternal)>
                    </div>
                  </StatusInParagraph>
                </FailedStatus>
              </div>
            </SingleReportStatus>
          </li>
          <li
            id="sonar"
            key="sonar"
          >
            <SingleReportStatus
              report={
                Object {
                  "name": "sonar",
                  "status": "PASS",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "flex",
                    "justifyContent": "space-between",
                    "width": "10rem",
                  }
                }
              >
                <span
                  id="report-name"
                >
                  sonar
                </span>
                 
                <PassedStatus>
                  <StatusInParagraph
                    color="green"
                    text="Passed"
                  >
                    <div
                      style={
                        Object {
                          "width": "4em",
                        }
                      }
                    >
                      <WithAnalyticsEvents(StatusInternal)
                        color="green"
                        text="Passed"
                      >
                        <AnalyticsContextConsumer
                          createEventMap={
                            Object {
                              "onClick": [Function],
                              "onHover": [Function],
                            }
                          }
                          wrappedComponentProps={
                            Object {
                              "color": "green",
                              "text": "Passed",
                            }
                          }
                        >
                          <StatusInternal
                            color="green"
                            createAnalyticsEvent={[Function]}
                            onClick={[Function]}
                            onHover={[Function]}
                            text="Passed"
                          >
                            <span
                              className="status-lozenge-span"
                              data-color="green"
                              data-node-type="status"
                              onClick={[Function]}
                              onMouseEnter={[Function]}
                              onMouseLeave={[Function]}
                              title="Passed"
                            >
                              <Lozenge
                                appearance="success"
                                isBold={false}
                                maxWidth={200}
                              >
                                <Provider>
                                  <Consumer
                                    appearance="success"
                                    isBold={false}
                                    maxWidth={200}
                                  >
                                    <Component
                                      backgroundColor="#E3FCEF"
                                      maxWidth={200}
                                      textColor="#006644"
                                    >
                                      <EmotionCssPropInternal
                                        __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Component"
                                        __EMOTION_TYPE_PLEASE_DO_NOT_USE__="span"
                                        css={
                                          Object {
                                            "backgroundColor": "#E3FCEF",
                                            "borderRadius": "3px",
                                            "boxSizing": "border-box",
                                            "color": "#006644",
                                            "display": "inline-block",
                                            "fontSize": "11px",
                                            "fontWeight": 700,
                                            "lineHeight": 1,
                                            "maxWidth": "100%",
                                            "padding": "2px 0 3px 0",
                                            "textTransform": "uppercase",
                                            "verticalAlign": "baseline",
                                          }
                                        }
                                      >
                                        <span
                                          className="css-1d02ga6-Component"
                                        >
                                          <Component
                                            backgroundColor="#E3FCEF"
                                            maxWidth={200}
                                            textColor="#006644"
                                          >
                                            <EmotionCssPropInternal
                                              __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Component"
                                              __EMOTION_TYPE_PLEASE_DO_NOT_USE__="span"
                                              css={
                                                Object {
                                                  "boxSizing": "border-box",
                                                  "display": "inline-block",
                                                  "maxWidth": "200px",
                                                  "overflow": "hidden",
                                                  "padding": "0 4px",
                                                  "textOverflow": "ellipsis",
                                                  "verticalAlign": "top",
                                                  "whiteSpace": "nowrap",
                                                  "width": "100%",
                                                }
                                              }
                                            >
                                              <span
                                                className="css-19vb09t-Component"
                                              >
                                                Passed
                                              </span>
                                            </EmotionCssPropInternal>
                                          </Component>
                                        </span>
                                      </EmotionCssPropInternal>
                                    </Component>
                                  </Consumer>
                                </Provider>
                              </Lozenge>
                            </span>
                          </StatusInternal>
                        </AnalyticsContextConsumer>
                      </WithAnalyticsEvents(StatusInternal)>
                    </div>
                  </StatusInParagraph>
                </PassedStatus>
              </div>
            </SingleReportStatus>
          </li>
        </ul>
      </div>
    </ReportStatus>
    <OverviewAdmonition
      userAdmin={false}
    >
      <div
        id="overview-admonition"
      />
    </OverviewAdmonition>
  </div>
</OverviewModal>
`;
